src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers/ApiControllerAnalyzer.cs

```diff
@@ -0,0 +1,110 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers
+{
+    [ApplicableComponents(ProjectComponents.AspNetCore)]
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class ApiControllerAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "UA0013";
+        private const string Category = "Upgrade";
+
+        public const string ApiControllerQualifiedName = ApiControllerNamespace + "." + ApiControllerClassName;
+        public const string ApiControllerNamespace = "System.Web.Http";
+        public const string ApiControllerClassName = "ApiController";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.ApiControllerTitle), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.ApiControllerMessageFormat), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.ApiControllerDescription), Resources.ResourceManager, typeof(Resources));
+
+        protected static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            // noted that highlighting the syntax makes it easier to handle partial class scenarios because
+            // the analyzer will be limited to scenarios where a base type has been defined via syntax
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, CodeAnalysis.CSharp.SyntaxKind.BaseList);
+            context.RegisterSyntaxNodeAction(AnalyzeSyntax, CodeAnalysis.VisualBasic.SyntaxKind.InheritsStatement);
+        }
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var node = context.Node; // an ImportStatementSyntax for VB or a BaseListSyntax for CS
+
+            if (!node.DescendantNodes().Any())
+            {
+                return;
+            }
+
+            // the first descendent of ImportStatementSyntax will be a QualifiedNameSyntax or IdentifierNameSyntax for VB
+            // the first descendent of BaseListSyntax will be a SimpleBaseTypeSyntax for CS
+            var baseTypeNode = context.Node.DescendantNodes().First();
```

Comment:
Yes, we could have used a DescendentNodes.OfType filter.

I felt the flow looked simpler without additional C# and VB specific syntax noise. LMK if you want to get specific about this.

Reply:
made more sense as an approach when I considered the fragility of depending on a specific node tree and the possibility of abstracting away language syntax to extension methods that focused on intent rather than syntax.